### PR TITLE
Use CaretTargets to handle deleting selection

### LINKED ISSUE #4264: Off-by-one Tuple Reconstruction bug
Given the tuple, `("aaa|aa", "c|cccc", "eeeee")` highlighting between the `|`s and cutting should leave `("aaa", "eeeee")` but leaves `("aaaa", "eeeee")`. There's some off-by-one error here.

Once these are done, 2 TUPLETODO commented-out tests are likely to auto-resolve.

### PR DESCRIPTION
**Pull request by StachuDotNet**
Changelog:

```
Editor improvements:
- deleting a selected range of code is now more accurate
```

This will fix #4264

Our process to delete a "selection" of Dark code is currently pretty naive, and we're trying to make it much more solid and correct.

Currently, `Fluid.deleteCaretRange` deletes from the end of the selection to the start of the selection character by character.
In order to determine when to stop deleting, it checks for:
- is the 'current' position `<=` the selection's start pos? If so, we're done.
- has the last 'delete' done anything (changed our pos or changed the AST)? if not, stop, so we don't loop forever.

This falls short in situations where the position changes throughout the deletion process.
For example, consider if you start with the AST `(123, 456)` and select `(12|3, 4|56)`. Throughout the deletion (upon deleting the `,`), the second part of the tuple is deleted, which results in the tuple being ripped out, replacing the tuple with a simple `123`. When this happens, the position of `12|3` is suddenly `2` where it used to be `3`, and we end up deleting too little.

Here's a showcase of the bug:
old

To resolve this, we should use CaretTargets - they refer to offsets of tokenizations of AST parts, and should be generally stable throughout the deletion loop.

With the fix, this is what happens when you `cut` in the same scenario:

new

---

It was tricky to ensure we had no regressions, but there are many tests around Fluid reconstruction which steered this away from a few wrong paths

### CODE REVIEW

File: client/test/TestFluid.res [hunk 1]
```diff
@@ -2852,7 +2852,7 @@ let run = () => {
       aConstructor,
       ~pos=4,
       inputs(list{DeleteWordForward}),
-      "Just ___~",
+      "Just ~___",
```
Comment (StachuDotNet) [inline client/test/TestFluid.res:2855]: I'm not sure how, but this improvement came for free - the cursor is now at the start of the blank rather than the end.

File: client/test/TestFluidClipboard.res [hunk 2]
```diff
@@ -951,7 +951,7 @@ let run = () => {
       if'(bool(true), str("then body"), str("else body")),
       (12, 31),
       (
-        "if true\nthen~\n  ___\nelse\n  \"else body\"",
+        "if true\nthen\n  ~___\nelse\n  \"else body\"",
```
Comment (StachuDotNet) [inline client/test/TestFluidClipboard.res:954]: I'm not sure how, but this improvement came for free.

File: client/src/fluid/Fluid.res [hunk 3]
```diff
@@ -1000,6 +1008,35 @@ let caretTargetForNextNonWhitespaceToken = (~pos, tokens: tokenInfos): option<CT
   getNextWS(tokens)
 }
 
+@ocaml.doc("Gets the caret target for a pos within a list of tokens. Ignores
+  whitespace tokens, and favors tokens 'to the left' of the pos, in cases where two
+  tokens surround the position.")
+let caretTargetForCurrentNonWhitespaceToken = (~pos, tokens: tokenInfos): option<CT.t> =>
+  List.fold(
```
Comment (pbiggar) [inline client/src/fluid/Fluid.res:1015]: I'm not 100% sure, but it feels like this could be a `List.find` instead of a `fold`

**Review by pbiggar — CHANGES_REQUESTED** [review]
I really like this new algorithm, makes a lot of sense to me.

However, I think there's a bug in the changes to DeleteWordForward - I feel that that has an invariant that the cursor stays in the same place. Perhaps we need to ensure that the cursor is put back at the place it starts? Although I haven't looked out how DeleteWordForward hooks into this.

Would also like a few more tests to hit some more edge cases - mostly where is the cursor before and after if it's on different tokens. Or do you think that's well covered already?

**Comment by StachuDotNet** [comment]
I've pushed a commit to revert those tests back (so they match what you expect).
I have mixed feelings about it.

While in the loop, the adjustment here is to:
if the "current" astref matches the 'start' astRef, keep backspacing until we're either "before" the starting position, or the 'current' astref no longer matches the 'start' astref

This works great for all tests, but feels a bit odd. Over all, the solution feels better than what we started with, but still not ideal.
Specifically, I'm concerned about what happens if the 'starting position' adjusts a bit (e.g. by removing the wrapping structure).

My next step here will be to create a test deleting the selection shown in this expr:
```
(if true
 then
    "then body"
 else|
    "else body",tr|ue)
```

I suspect it will break.

tricky...

File: client/test/TestFluidClipboard.res [hunk 4]
```diff
@@ -1233,12 +1245,12 @@ let run = () => {
...
-    //   tuple(str("hello"), int(1234), list{}), // ("hello",1234)
-    //   (5, 11),
-    //   ("(\"hel~\")", "(\"lo\",12)"),
-    // )
+    testCut(
+      "cutting halfway between tuple parts leaves a partial tuple and copies the data selected to clipboard",
+      tuple(str("hello"), int(1234), list{}), // ("hello",1234)
+      (5, 11),
+      ("\"hel~\"", "(\"lo\",12)"),
+    )
```
Comment (StachuDotNet) [inline client/test/TestFluidClipboard.res:1253]: This was the test that this PR was actually aimed at.

File: client/test/TestFluidClipboard.res [hunk 5]
```diff
@@ -955,6 +955,25 @@ let run = () => {
...
+    // deleting early.
+    // Initially, the plan is to delete until '4 characters after the end of the `else`'
+    // But with the tuple being removed, we want to end at '3 chars after the end of the `else`',
+    // and currently have no good way to 'learn' this along the way.
+    // testCut(
+    //   "same thing as above but we end just before the else branch's string expr",,
+    //   tuple(if'(bool(true), str("then body"), str("else body")), bool(true), list{}),
+    //   (39, 53), // from `|"else body"` to `tr|ue`
+    //   ("if true\nthen\n  \"then body\"\nelse\n  ~___", "(\"else body\",tr)"),
+    // )
```
Comment (StachuDotNet) [inline client/test/TestFluidClipboard.res:976]: Note: this scenario behaves the same as it did before this PR.
This test was a stretch goal of the PR that I couldn't quite figure out at this point.

File: client/src/fluid/Fluid.res (hunk 3 above)
Comment (StachuDotNet) [reply client/src/fluid/Fluid.res:1015]: I couldn't figure out how :) Open to suggestions!

**Comment by StachuDotNet** [comment]
> Would also like a few more tests to hit some more edge cases - mostly where is the cursor before and after if it's on different tokens. Or do you think that's well covered already?

I believe we're pretty well covered, though I'm happy to add tests if you have ideas?

**Comment by StachuDotNet** [comment]
I've gone to tackle this algorithm in a bunch of different ways at this point, and have sort of exhausted my brain on the topic for now. There are scenarios I wanted to solve for as part of this (see the newly-added and commented test), but I'm satisfied following up on those later.

I think we're pretty well covered, but would be happy to add additional tests if you had ideas?

**Comment by StachuDotNet** [comment]
To be really clear, the scenario I want to resolve, but hasn't yet been resolved, is this:

```
(if true
 then
    "then body"
 else
    |"else body",tr|ue)
```

The start of the selection is "4 characters after the _end_ of the `else` keyword's ASTRef" when we start "hitting backspace."
As we 'hit backspace' in the loop, the indentation of the `else` is removed once the surrounding tuple is gone. And so, we end up backspacing too little - the `"else body"` that we expect to be removed in favor of a `blank ___` is instead left as an empty string with our cursor in it (`"|"`), as _that_ is now 4 characters after the end of the `else`'s ASTRef.

**Comment by StachuDotNet** [comment]
worth considering next time we tackle this:
"stop backspacing when we're at the end of the whitespace following the caret token"
and variants of such.

(rather than using an 'offset' from a caret target)

**Review by pbiggar — APPROVED** [review]
Looks good, but needs a changelog in the description.
